DataScience/Analytics/prototype/automatic_testing.py

```diff
@@ -0,0 +1,254 @@
+    for i, method1 in enumerate(methods):
+        for j, method2 in enumerate(methods):
+            if method1 == 'UPRN_SAS' and method2 == 'UPRN_PROTO':
+                agree = df[method1] == df[method2]
+                nagree = len(df.loc[agree].index)
+
+                msk = (~df[method1].isnull()) & (~df[method2].isnull())
+                disagree = df.loc[msk, method1] != df.loc[msk, method2]
+                ndisagree = len(df.loc[msk & disagree].index)
+
+                msk = (df[method1].isnull()) & (~df[method2].isnull())
+                nmethod2only = len(df.loc[msk].index)
+
+                correct = nagree
+                false_positive = ndisagree
+                new_UPRNs = nmethod2only
+
+    results = dict(addresses=addresses, correct=correct, false_positive=false_positive, linked=linked,
+                   new_UPRNs=new_UPRNs, not_linked=not_linked, withUPRN=withUPRN)
+
+    return results
+
+
+def compute_performance():
+    """
+    Computes additional performance metrics as some datasets have multiple UPRNs attached or
+    UPRNs have been attached later.
+
+    :return: None
+    """
+    welsh_data = _load_welsh_data()
+
+    # compute results and create a dictionary
+    results = _compute_welsh_performance(welsh_data, methods=('UPRN_PROTO', 'UPRN_SAS'))
+    results['code_version'] = addressLinking.__version__
+    results['dataset'] = 'WelshGovernmentData21Nov2016.csv'
+    results['date'] = datetime.datetime.now()
+    results['name'] = 'WelshGovSAS'
+
+    # convert to Pandas Dataframe
+    results = pd.DataFrame.from_records([results])
+
+    # push to the database
+    with sqlite3.connect(location + 'AddressLinkingResults.sqlite') as cnx:
+        results.to_sql('results', cnx, index=False, if_exists='append')
+
+
+def _get_data_from_db(sql):
+    """
+    Pull data from a database.
+
+    :param sql: sql query to execute to pull the data
+    :type sql: str
+
+    :return: queried data
+    :rtype: pandas.DataFrame
+    """
+    # build the connection string from specifying the DB type, location, and filename separately
+    connection = 'sqlite:///' + location + 'AddressLinkingResults.sqlite'
+
+    df = pd.read_sql_query(sql, create_engine(connection))
+
+    return df
+
+
+def plot_performance():
+    """
+    Generates simple graphs which show the linking performance as a function of time for all datasets
+    available from the results database.
+
+    For each dataset two graphs are generated: 1) figure with multiple sub-figures, and 2)
+    a single figure showing multiple lines.
+
+    :return: None
+    """
+    # query data and place it to a Pandas DataFrame
+    data = _get_data_from_db('select * from results;')
+
+    # convert date to datetime
+    data['date'] = pd.to_datetime(data['date'])
+
+    columns_to_plot = ['addresses', 'correct', 'false_positive', 'linked', 'new_UPRNs', 'not_linked']
+
+    # create figures
```

Comment:
This is a kind of comment that would better be a private function named `_createFigures`